### DISCUSSION TITLE
GetServerSideProps returns undefined in Pages file

### DISCUSSION THREAD

**Discussion by molikkidd** [vercel/next.js#33509] [Help]
I am calling from a simple api and looking to pass the response data as a property to my page. I have read many stackOverFlow docs stating that you cannot use getServerSideProps from a component so I implemented it a page/ as suggested.  The API works and is returning data, I confirmed it in the console but when I pass on the data to the component through the suggested way, I get undefined. I tried (JSON.stringify(data)) and it didnt resolve the issue, I still get undefined. Any help will be great. 

```jsx
 export async function getServerSideProps(context) {

    const res = await fetch('http://localhost:8000/api/appQuestions/');
    const qData = await res.json();

    console.log("ServerSide call Data", qData);
    
    return {
      props: {
          questions: qData,
        }, // will be passed to the page component as props
    }
  }

const newapp = ({questions}) => {

    console.log("api data inside component newapp", questions)
    return (
        <div>
            <h1>hello moto</h1> )
}

```

**Comment by leerob** [comment]
Rather than fetching from `/api/appQuestions`, you can put that logic directly inside `getServerSideProps` - both run on the server. This will save an additional API call.

**Reply by molikkidd** [reply]
Can you show me an example because I thought that is what is in my code snippet? I was able to use a work around to get the data to populate the page without getServerSideProps, I used useEffect(), but after further research, they suggested against using an async function from within a component, even though its on a page route.

**Reply by leerob** [reply]
Whatever code is inside of the file `/api/appQuestions/`, move that directly inside `getServerSideProps` 😄

**Reply by molikkidd** [reply]
I dont think that will solve the issue of my props returning "undefined" in the component because the getServerSideProps already returns the data I need it to.

**Reply by nhnasem** [reply]
I am having the same issue with it. I didn't do res.json() and console logged the res. It returns undefined. Have you resolved the issue yet?

**Reply by HarshadSahu** [reply]
I have the same issue, can anyone help?

**Reply by nhnasem** [reply]
I resolved my issue by removing some errors from my code. For example, I was importing the Model incorrectly and destructured the data that I needed and used it as props. For the above code, doing `const { data } = await res.json()` instead of `const data = await res.json()` fixed my issue. If it still doesn't fix the problem, then the problem is elsewhere.

**Reply by HarshadSahu** [reply]
Thanks for the reply @nhnasem , yes you were right the problem was not in an import statement, it was something else. I'd resolve the issue from my end. Thanks.

**Reply by Dami-Olus** [reply]
I am having the same issue, @HarshadSahu how did you resolve your issue

**Reply by ZeddZoul** [reply]
If i console.log data..  it returns the expected object but when i try to access from page component
it returns undefined

By the way my page component is a dynamic id page

**Reply by faiz-9** [reply]
Did you find any solution mate?

**Reply by hassanah07** [reply]
Nextjs is now uses server component by default. If you use "use client" The aap will recognize as CSR. Use fetch api without "use client"

**Comment by hassanah07** [comment]
Hello Guys, I am trying to get data  as given solution above but still i am getting undefined instead of data but when I send this same request through postman and thunderclient  it gives me success with 200 status code. Please Guys, help me out...
```
 export async function getServerSideProps(context) {

    const res = await fetch(`http://localhost:5000/api/blogpost/view`);
    const { data } = await res.json();
    return {
      props: {
          blog: data,
        },
    }
  }
const Page = ({blog}) => {
    console.log({ blog: blog, title: "nothing" });
    return (
        <div><h1>Hello</h3></div>
 )}

export default Page;
```
### **I have tried the fetch request inside try-catch and then also I got same undefined message** 

```
export async function getServerSideProps(context) {
  // Fetch data from external API
  try {
    const res = await fetch(`http://localhost:5000/api/blogpost/view`);
    const { data } = await res.json();
  } catch (error) {
    console.log(error);
  }
  console.log(data);

  // Pass data to the page via props
  return { props: { blog: data } };
}
```
### **Console Data that I am getting**
```
{ blog: undefined, title: 'nothing' }
```

**Reply by icyJoseph** [reply]
Do you have an _app file? is it passing down pageProps to the Component?

**Reply by hassanah07** [reply]
No, I do not have _app file. I think the previous version of nextjs comes
with _app file.  I am sharing my files and folder structure
![Screenshot 2023-08-01 172159](https://github.com/vercel/next.js/assets/68686728/96994c4a-c0bd-41e0-bad1-01d0228fb5cf)

**Reply by icyJoseph** [reply]
In the App Router there's no `getServerSideProps`. You have to do the data fetching yourself. https://github.com/vercel/next.js/discussions/53294

**Reply by hassanah07** [reply]
How can I can do it my friend?

**Reply by icyJoseph** [reply]
Look into the link I posted: https://github.com/vercel/next.js/discussions/53294#discussioncomment-6586443
